felix/vxlanfdb/vxlan_fdb.go

```diff
@@ -0,0 +1,351 @@
+		// kernel that, if it needs to send traffic to the VTEP MAC, it
+		// should send the VXLAN packet to a particular host's real IP
+		// address.
+		f.fdbEntries.Desired().Set(macStr, ipMACMapping{
+			MAC: t.TunnelMAC,
+			IP:  t.HostIP,
+		})
+	}
+}
+
+func (f *VXLANFDB) Apply() error {
+	nl, err := f.nl.Handle()
+	if err != nil {
+		return fmt.Errorf("failed to connect to netlink")
+	}
+
+	if f.resyncPending {
+		if err := f.resync(nl); err != nil {
+			return err
+		}
+		f.resyncPending = false
+	}
+	defer func() {
+		if f.resyncPending {
+			f.logNextSuccess = true
+		}
+	}()
+
+	if f.ifIndex == 0 {
+		return ErrLinkDown
+	}
+
+	f.applyFamily(nl, "ARP/NDP", f.arpEntries,
+		func(mapping ipMACMapping) *netlink.Neigh {
+			return &netlink.Neigh{
+				Family:       f.family,
+				LinkIndex:    f.ifIndex,
+				State:        netlink.NUD_PERMANENT,
+				Type:         unix.RTN_UNICAST,
+				IP:           mapping.IP.AsNetIP(),
+				HardwareAddr: mapping.MAC,
+			}
+		},
+	)
+
+	f.applyFamily(nl, "FDB", f.fdbEntries,
+		func(mapping ipMACMapping) *netlink.Neigh {
+			return &netlink.Neigh{
+				Family:       unix.AF_BRIDGE,
+				LinkIndex:    f.ifIndex,
+				State:        netlink.NUD_PERMANENT,
+				Flags:        netlink.NTF_SELF,
+				IP:           mapping.IP.AsNetIP(),
+				HardwareAddr: mapping.MAC,
+			}
+		},
+	)
+
+	if !f.resyncPending && f.logNextSuccess {
+		f.logCxt.Info("VXLAN FDB now in sync.")
+		f.logNextSuccess = false
+	}
+	if f.resyncPending {
+		return fmt.Errorf("failed to add/delete some neighbor entries")
+	}
+
+	return nil
+}
+
+func (f *VXLANFDB) applyFamily(
+	nl netlinkshim.Interface,
+	description string,
+	entries *deltatracker.DeltaTracker[string, ipMACMapping],
+	entryToNeigh func(mapping ipMACMapping) *netlink.Neigh,
+) {
+	debug := log.IsLevelEnabled(log.DebugLevel)
+	errs := map[string]error{}
+	entries.PendingUpdates().Iter(func(macStr string, entry ipMACMapping) deltatracker.IterAction {
+		if debug {
+			log.WithField("entry", entry).Debugf("Adding %s entry.", description)
+		}
+		neigh := entryToNeigh(entry)
+		if err := nl.NeighSet(neigh); err != nil {
+			if len(errs) == 0 {
+				log.WithError(err).WithField("entry", entry).Warnf("Failed to add %s entry, only logging first instance.", description)
+			}
+			errs[macStr] = err
+			return deltatracker.IterActionNoOp
+		}
+		return deltatracker.IterActionUpdateDataplane
+	})
+	if len(errs) > 0 {
+		log.WithField("numErrors", len(errs)).Warnf("Failed to add some %s entries", description)
+		f.resyncPending = true
+		f.nl.MarkHandleForReopen() // Defensive: force a netlink reconnection next time.
+		clear(errs)
+	}
+
+	entries.PendingDeletions().Iter(func(macStr string) deltatracker.IterAction {
+		entry, _ := entries.Dataplane().Get(macStr)
+		if debug {
+			log.WithField("entry", entry).Debug("Deleting ARP entry.")
+		}
+		neigh := entryToNeigh(entry)
+		if err := nl.NeighDel(neigh); err != nil && !errors.Is(err, unix.ENOENT) {
```

Comment:
IIUC, here is where the point that you mentioned comes into play, about `NeighDel` on a constructed `neigh`, instead of on one returned by `NeighList`.  Am I right that you could eliminate this delta by passing the whole `neigh` as the value of the delta tracker, instead of `ipMacMapping`?  If so, I suppose the advantage of `ipMacMapping` is that you can use the same value for both trackers, whereas with the whole `neigh` you'd need two slightly different neighs.

I think the code is OK as it is; just commenting to check my understanding, and to point out the possibility in case it was appealing.